### ISSUE TITLE
BulkInsert and overwrite existing

### ISSUE THREAD

**Issue by jenspettersson** [JasperFx/marten#180]
In RavenDB there's a `BulkInsertOptions` where you can specify `OverwriteExisting`:

```
documentStore.BulkInsert(options: new BulkInsertOptions
{
    OverwriteExisting = true
});
```

In Marten the bulk insert is done by: `COPY mt_doc_test(\"data\", \"id\") FROM STDIN BINARY`

We talked a bit in the gitter chat about how and if this can be done against Postgres. There was a mention of a trigger `ON CONFLICT DO UPDATE` but there might be other ways of doing it? _(Note: My extremely limited Postgres 'skills' prevented me to create such a trigger and test locally...)_

We use the BulkInsert - OverwriteExisting quite a lot in a couple of projects where we replace most of the db every hour or so, but we don't want to truncate before as we want to keep older record that wasn't in the last load (we have timestamps for each document so we can see what entries is old/outdated).

Without the overwrite support we have to use the `session.Store(...)` for each document and let the UPSERT sort it out, but since we're processing millions of records this will cause a bit of an overhead.

Any ideas?

**Comment by jeremydmiller** [comment]
@jenspettersson I think we've got two options, we could either:

1. Automatically add the ON CONFLICT trigger to handle the conflicts
1. Make the bulk insert quietly switch to using the normal Unit of Work/batched update command instead of Postgresql's "COPY" functionality.

Both would incur a performance hit, but I don't know which one would be worse. The second option would be faster to code out.

**Comment by jenspettersson** [comment]
@jeremydmiller If I can get the ON CONFLICT trigger to work I can do some testing locally using both approaches (adding the trigger manually that is). Any tips on how the trigger could look like?

**Comment by jeremydmiller** [comment]
I struck out googling for that, and I think I'm mistaken because I don't see any option for ON CONFLICT for trigger definitions in the postgresql docs.

**Comment by jenspettersson** [comment]
That explains... a lot! I started to doubt my googling skills there... =) Then there might not be another way than 2?

**Comment by jamesfarrer** [comment]
I had a quick look / google and there doesn't seem to be a clean way to do this with the built in COPY tool or a conflict trigger.

I don't know what you would think of this solution, but my idea would be that if overwrite existing is true :-

- Create a new (temporary) table and execute the COPY into it
- Insert any id's that don't exist into the primary table
- Update any id's that exist in the primary table
- Drop the new (temporary) table

This would at least allow for some bulk INSERT / UPDATE operations which should be pretty fast

Hopefully this would only be temporary until someone improves the COPY util in pgSQL to support overwrite

**Comment by jenspettersson** [comment]
As I said in the gitter chat, I'm not a PG expert by far, but if the solution @jamesfarrer suggest is doable in a clean and secure way, then why not. It does feel a bit "Rube Goldberg" however ;)

I'm not in a position to call any shots here, so we might need @jeremydmiller to comment as well.

**Comment by jeremydmiller** [comment]
Throwing this link in here: http://stackoverflow.com/questions/30500560/import-and-overwrite-duplicate-rows. 

@jenspettersson @jamesfarrer Do we want to throw away the duplicates, or overwrite? I suppose that it could be done with a mode like "Fast" / "IgnoreDuplicates" / "Overwrite"

**Comment by jenspettersson** [comment]
It might be my severe "man cold" that's making me numb, but what's the difference between "throwing away" and "Overwrite" a duplicate?

Do you mean: First delete the duplicate and then insert a new instead of just overwriting it? I don't see a need for that, but then again, I think I'm missing something...

**Comment by jeremydmiller** [comment]
Ignore duplicate values altogether vs. last one in wins

**Comment by jenspettersson** [comment]
Ah you mean if there's a duplicate within the new batch that are to be bulk inserted. Yeah that also need to be handled. In all my cases I have logic that handles duplicates within current batch before doing a bulk insert so I've never thought about this.

**Comment by jamesfarrer** [comment]
Yup I think the 3 config options make sense. For overwrite I think an update would be preferable. Deleting could cause potential FK exceptions specified on that document / table.

So for the 3 options.. 

Fast (Default) = standard copy as used today

IgnoreDuplicates would be

create temporary table import_issues as select * from issues limit 0; (I've cribbed this from that SO post but assume as we'd have the schema definition we'd use that)

copy import_issues from stdin;
begin transaction

insert into issues
select * from import_issues where id not in (select id from issues)
end transaction

Overwrite would be

create temporary table import_issues as select * from issues limit 0; (I've cribbed this from that SO post but assume as we'd have the schema definition we'd use that)

copy import_issues from stdin;
begin transaction

insert into issues
select * from import_issues where id not in (select id from issues)

update i set i.data = i2.data
from issues i
inner join import_issues i2 on i.id = i2.id
end transaction
